src/core/libmaven/PeakDetector.cpp

```diff
@@ -376,103 +527,435 @@ void PeakDetector::processSlices(vector<mzSlice*> &slices, string setName)
-                    && groupWithTarget.ms2EventCount > 0
-                    && groupFiltering.filterByMS2(groupWithTarget)) {
-                    continue;
-                }
+            if (!orphans.empty()) {
+                // for orphans, create a ghost, that will act as an empty parent
+                PeakGroup parentGroup(
+                    make_shared<MavenParameters>(*_mavenParameters),
+                    PeakGroup::IntegrationType::Ghost);
+                container.push_back(parentGroup);
+
+                // set an appropriate slice for ghost parent
+                mzSlice slice;
+                slice.compound = compound;
+                slice.calculateMzMinMax(
+                    _mavenParameters->compoundMassCutoffWindow,
+                    _mavenParameters->getCharge(compound));
+                slice.calculateRTMinMax(false, 0.0f);
+                container.back().setSlice(slice);
+
+                size_t totalSize = container.size();
+                container.back().setGroupId(totalSize);
+                metaGroups[totalSize - 1] = {};
+                for (auto child : orphans)
+                    metaGroups[totalSize - 1].push_back(child);
+            }
 
-                matchFound = true;
-                toBeMerged.push_back(groupWithTarget);
+            return metaGroups;
+        };
+
+    unordered_map<Compound*, unordered_map<size_t, vector<size_t>>> metaGroups;
+
+    // find isotope meta-groups
+    for (auto& elem : nonParentIsotopologues) {
+        if (_mavenParameters->stop)
+            return;
+
+        Compound* compound = elem.first;
+        auto& isotopeIndexes = elem.second;
+        auto metaIsotopeGroups = makeMeta(compound,
+                                          isotopeIndexes,
+                                          parentCompounds,
+                                          [](PeakGroup* group) {
+                                              return group->isotope().name;
+                                          });
+        metaGroups[compound] = metaIsotopeGroups;
+    }
+
+    // find adduct meta-groups
+    for (auto& elem : nonParentAdducts) {
+        if (_mavenParameters->stop)
+            return;
+
+        Compound* compound = elem.first;
+        auto& adductIndexes = elem.second;
+        auto metaAdductGroups = makeMeta(compound,
+                                         adductIndexes,
+                                         parentCompounds,
+                                         [](PeakGroup* group) {
+                                             return group->adduct()->getName();
+                                         });
+        if (metaGroups.count(compound) > 0) {
+            auto& existingMetaGroups = metaGroups[compound];
+            for (auto& elem : metaAdductGroups) {
+                size_t parentIndex = elem.first;
+                vector<size_t>& childIndexes = elem.second;
```

Comment:
Would suggest renaming `childIndexes` to `adductIndices` or `childAdductIndices` for clarity. Hard to read this block with so many similar variable names.